flow-server/src/main/java/com/vaadin/flow/server/VaadinSession.java

```
@@ -145,6 +140,14 @@ public class VaadinSession implements HttpSessionBindingListener, Serializable {
      */
     public VaadinSession(VaadinService service) {
         this.service = service;
+
+        if (service != null) {
```

I wonder if we should throw an exception when service is null.
A VaadinSession without a service reference will not work correctly, and the only moment I can imagine the field to be null is on deserialization, but in that case, the constructor will not be invoked.

---

You are right, that would make sense. I'll do the change, although it will mean revisiting at least one (quite old) test, which is calling the constructor with null parameter (claiming it as an 'uninitialized' session, whatever that means).

---

Could it be a post deserialization simulation? Anyway, I would probably nullify the field through reflection in that case

---

It's this one: https://github.com/vaadin/flow/blob/main/flow-server/src/test/java/com/vaadin/flow/server/VaadinSessionTest.java#L483

---

HTTP session unboud is called when the implementing object is removed from the HTTP session, so I think that in that case the VaadinSession should have a not-null reference to VaadinService

---

Ok, I'll have to check how to fix it then. Initially I just changed the null to a mock VaadinService but that will fail the test.

---

Ah, my bad. The `valueUnbound` method Javadoc explains the case.
I propose to have the null-check in the constructor and nullify it with reflection (or serialize/deserialized the VaadinSession object)

---

Fixed